File: src/Y2024/Day19/main.rb
designs = ""
towels = []
File.open("input.txt") { |f|
  designs, _, *towels = f.map(&:chomp)
}
trie = {marked: false}
designs.split(", ") { |design|
  node = trie
  design.each_char { |s|
    if !node.key?(s)
      node[s] = {marked: false}
    end

    node = node[s]
  }
  node[:marked] = true
}

def count(trie, towel)
  cache = {}
  tl = towel.length
  solve = lambda do |idx|
    if idx == tl
      return 1
    elsif cache.key?(idx)
      return cache[idx]
    end

    count = 0
    node = trie
    (idx..tl).each do |i|
      s = towel[i]
      if !node.key?(s)
        break
      end

      node = node[s]
      if node[:marked]
        count += solve.call(i + 1)
      end
    end

    cache[idx] = count
  end

  solve.call(0)
end

p1 = p2 = 0
towels.each do |towel|
  res = count(trie, towel)
  if res > 0
    p1 += 1
  end

  p2 += res
end

puts("#{p1} #{p2}")

File: src/Y2024/Day19/main.py
from typing import TextIO, Self
from attrs import define, field
from functools import cache

from Tools.Python.path_stuff import *
from Tools.Python.utils.utils import Res, res_sum


def parser(raw_data: TextIO):
    designs, towels = raw_data.read().split('\n\n')
    trie = Trie()
    for design in designs.split(', '):
        node = trie
        for s in design:
            node = node[s]
        node.marked = True
    return trie, towels.splitlines()


@define
class Trie:
    children: dict[str, Self] = field(factory=dict)
    marked: bool = False

    def __getitem__(self, key):
        if key in self.children:
            return self.children[key]
        self.children[key] = Trie()
        return self.children[key]


def check(designs: Trie, towel: str):
    tl = len(towel)

    @cache
    def solve(idx: int):
        if idx == tl:
            return 1
        node = designs
        count = 0
        for i in range(idx, tl):
            s = towel[i]
            if s not in node.children:
                break
            node = node.children[s]
            if node.marked:
                count += solve(i + 1)
        return count
    return solve(0)


def full_solver(designs: Trie, towels: list[str]):
    return res_sum(
        Res(1, res) for towel in towels
        if (res := check(designs, towel))
    )


if __name__ == '__main__':
    testing = False

    with open(test_path if testing else 'input.txt', 'r') as file:
        data = parser(file)

    print(*full_solver(*data))
